File: src/main/java/com/nanaios/polygonal_tech/util/IOMode.java
package com.nanaios.polygonal_tech.util;

import com.nanaios.polygonal_tech.util.interfaces.IIOMode;

public class IOMode {
    /// デフォルトのINPUTモード。通常は、入力が許可されることを示す。
    public static final IIOMode INPUT = new Impl("input");
    /// デフォルトのOUTPUTモード。通常は、出力が許可されることを示す。
    public static final IIOMode OUTPUT = new Impl("output");
    /// デフォルトのINPUT_OUTPUTモード。通常は、入出力の両方が許可されることを示す。
    public static final IIOMode INPUT_OUTPUT = new Impl("input_output");
    /// デフォルトのNONEモード。通常は、入出力を許可しないことを示す。
    public static final IIOMode NONE = new Impl("none");

    /// IOModeの単純な実装クラス。
    public record Impl(String name) implements IIOMode { }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/lib/util/IOMode.kt
package com.nanaios.polygonal_tech.lib.util

import com.nanaios.polygonal_tech.main.PolygonalTech
import net.minecraft.resources.ResourceLocation

interface IIOMode {
    val id: ResourceLocation
}

enum class IOMode(
    override val id: ResourceLocation,
    val canInput:Boolean,
    val canOutput:Boolean,
): IIOMode {
    NONE(createIOModeLocation(PolygonalTech.MOD_ID,"none"), false, false),
    INPUT(createIOModeLocation(PolygonalTech.MOD_ID,"input"), true, false),
    OUTPUT(createIOModeLocation(PolygonalTech.MOD_ID,"output"), false, true),
    INPUT_OUTPUT(createIOModeLocation(PolygonalTech.MOD_ID,"input_output"), true, true)
}

fun createIOModeLocation(modId: String,name:String): ResourceLocation {
    return ResourceLocation.fromNamespaceAndPath(modId, "io_mode.$name")
}